provider/packet/provider.go

```diff
@@ -0,0 +1,137 @@
+package packet
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/juju/errors"
+	"github.com/juju/jsonschema"
+	"github.com/juju/juju/environs"
+	environscloudspec "github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/context"
+	"github.com/lxc/lxd/shared/logger"
+	"github.com/packethost/packngo"
+)
+
+type environProvider struct {
+	environProviderCredentials
+}
+
+func (p environProvider) CloudSchema() *jsonschema.Schema {
+	trace()
+	return nil
+}
+
+// Ping tests the connection to the cloud, to verify the endpoint is valid.
+func (p environProvider) Ping(ctx context.ProviderCallContext, endpoint string) error {
+	trace()
+	return errors.NotImplementedf("Ping")
+}
+
+// PrepareConfig is part of the EnvironProvider interface.
+func (p environProvider) PrepareConfig(args environs.PrepareConfigParams) (*config.Config, error) {
+	trace()
+	if err := validateCloudSpec(args.Cloud); err != nil {
+		return nil, errors.Annotate(err, "validating cloud spec")
+	}
+	// Set the default block-storage source.
+	attrs := make(map[string]interface{})
+	// if _, ok := args.Config.StorageDefaultBlockSource(); !ok {
+	// 	attrs[config.StorageDefaultBlockSourceKey] = azureStorageProviderType
+	// }
+	// fmt.Println("config attrs", args.Config)
+
+	if len(attrs) == 0 {
+		return args.Config, nil
+	}
+	return args.Config, nil
+}
+
+func validateCloudSpec(spec environscloudspec.CloudSpec) error {
+	trace()
+	// if err := spec.Validate(); err != nil {
```

Comment:
This probably needs to be implemented

Reply:
I'm not sure what to validate in this case. Can you give me little more details?

Reply:
I believe you need to call `Validate` on the spec and ensure that the credential is not empty. Otherwise, if something goes wrong and the credential is not specified, you can get a panic when you try to access the credential bits (e.g. in `getPacketInstancesByTag`.